File: E-Tour/src/main/java/com/etour/controllers/AuthController.java
package com.etour.controllers;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.etour.dto.auth.AuthResponse;
import com.etour.dto.auth.ChangePasswordRequest;
import com.etour.dto.auth.LoginRequest;
import com.etour.dto.auth.RegisterRequest;
import com.etour.dto.auth.UpdateProfileRequest;
import com.etour.dto.auth.UserProfileDTO;
import com.etour.services.AuthService;

import jakarta.validation.Valid;

@RestController
@RequestMapping("/api/auth")
public class AuthController {
    private final AuthService authService;

    public AuthController(AuthService authService) {
        this.authService = authService;
    }

    @PostMapping("/register")
    public ResponseEntity<AuthResponse> register(@Valid @RequestBody RegisterRequest request) {
        return new ResponseEntity<>(authService.register(request), HttpStatus.CREATED);
    }

    @PostMapping("/login")
    public ResponseEntity<AuthResponse> login(@Valid @RequestBody LoginRequest request) {
        return new ResponseEntity<>(authService.login(request), HttpStatus.OK);
    }

    @GetMapping("/me")
    public ResponseEntity<UserProfileDTO> me(Authentication authentication) {
        return new ResponseEntity<>(
                authService.getProfile(authentication.getName()), HttpStatus.OK);
    }

    @PutMapping("/me")
    public ResponseEntity<UserProfileDTO> updateMe(
            @Valid @RequestBody UpdateProfileRequest request,
            Authentication authentication) {
        return new ResponseEntity<>(
                authService.updateProfile(authentication.getName(), request), HttpStatus.OK);
    }

    @PostMapping("/change-password")
    public ResponseEntity<java.util.Map<String, String>> changePassword(
            @Valid @RequestBody ChangePasswordRequest request,
            Authentication authentication) {
        authService.changePassword(authentication.getName(), request);

        return new ResponseEntity<>(
                java.util.Map.of("message", "Your password has been changed."), HttpStatus.OK);
    }
}

File: ETour-DotNet/Controllers/AuthController.cs
using ETour.Api.DTO;
using ETour.Api.Options;
using ETour.Api.Security;
using ETour.Api.Service;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Options;

namespace ETour.Api.Controllers;

[ApiController]
[Route("api/auth")]
public class AuthController : ControllerBase
{
    private readonly IAuthService _auth;
    private readonly OAuthOptions _oauth;

    public AuthController(IAuthService auth, IOptions<OAuthOptions> oauth)
    {
        _auth = auth;
        _oauth = oauth.Value;
    }

    [HttpPost("register")]
    public async Task<ActionResult<AuthResponse>> Register([FromBody] RegisterRequest request) =>
        StatusCode(StatusCodes.Status201Created, await _auth.RegisterAsync(request));

    [HttpPost("login")]
    public async Task<ActionResult<AuthResponse>> Login([FromBody] LoginRequest request) =>
        Ok(await _auth.LoginAsync(request));

    [Authorize]
    [HttpGet("me")]
    public async Task<ActionResult<UserProfileDTO>> Me() =>
        Ok(await _auth.GetProfileAsync(User.Username()));

    [Authorize]
    [HttpPut("me")]
    public async Task<ActionResult<UserProfileDTO>> UpdateMe(
        [FromBody] UpdateProfileRequest request) =>
        Ok(await _auth.UpdateProfileAsync(User.Username(), request));

    [Authorize]
    [HttpPost("change-password")]
    public async Task<ActionResult<Dictionary<string, string>>> ChangePassword(
        [FromBody] ChangePasswordRequest request)
    {
        await _auth.ChangePasswordAsync(User.Username(), request);
        return Ok(new Dictionary<string, string>
        {
            ["message"] = "Your password has been changed."
        });
    }

    [AllowAnonymous]
    [HttpGet("providers")]
    public ActionResult<Dictionary<string, object>> Providers() =>
        Ok(new Dictionary<string, object>
        {
            ["google"] = _oauth.IsConfigured(),
            ["googleAuthUrl"] = _oauth.IsConfigured() ? "/oauth2/authorization/google" : null
        });
}
